conjur/init_logic.py

```diff
@@ -0,0 +1,107 @@
+# Builtin
+import logging
+import sys
+import os.path
+from pathlib import Path
+
+# Third party
+import socket
+import yaml
+from OpenSSL import SSL
+
+class InitLogic:
+  """
+  InitLogic
+
+  This class holds the business logic for population the  conjurrc configuration details needed to connect with Conjur
+  """
+  DEFAULT_CONFIG_FILE = os.path.expanduser(os.path.join('~', '.conjurrc'))
+  DEFAULT_PROMPT_RESPONSE=True
+
+  def __init__(self):
+    pass
+
+  @staticmethod
+  def connect_to_socket(hostname, port):
+    """
+    Method for opening a socket with the Conjur server
+    """
+    try:
+      if port == None:
+        port = 443
+
+      ctx = SSL.Context(method=SSL.TLSv1_2_METHOD)
+      # We set the following to false because we want to fetch the certificates and not use them to validate yet
+      ctx.check_hostname = False
+      ctx.verify_mode = False
+
+      sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+      sock = SSL.Connection(context=ctx, socket=sock)
+      sock.connect((hostname, 443))
+      sock.do_handshake()
+    except Exception as e:
+      raise Exception(f"Unable to retrieve certificate from {hostname}:{port}")
+
+    logging.info("TLS Connection established. " \
+             "Fetching certificate from server...")
+    return sock
+
+  def get_certificate(self, hostname, port, sock):
+    """
+    Method for connecting to Conjur to fetch the certificate chain
+    """
+    from OpenSSL.crypto import dump_certificate, FILETYPE_PEM
+    try:
+      chain = sock.get_peer_cert_chain()
+      cert = chain[0]
+      fp = cert.digest("sha1").decode("utf-8")
+
+      # Get the peer chair and format to make it easier to later write to a file
+      readable_certificate=''
+      for (idx, cert) in enumerate(chain):
+          readable_certificate+=str(dump_certificate(FILETYPE_PEM, cert),  "utf-8")
+
+      logging.info("Certificates were fetched successfully")
+    except Exception:
+      raise Exception(f"Unable to retrieve certificate from {hostname}:{port}")
```

Comment:
Use ssl.CertificateError  or create your own exception for it.

Reply:
ssl.CertificateError doesn't apply here. I tried to change it here and in the greater try/catch in the outer `cli.py` and get `[Errno 8] nodename nor servname provided, or not known` so the execution here is a different. I will try to find the appropriate one